Backpack/StarRating/Classes/BPKStarRating.m

```diff
@@ -0,0 +1,147 @@
+ * limitations under the License.
+ */
+
+#import "BPKStarRating.h"
+
+#import <Backpack/Common.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface BPKStarRating ()
+
+@property (nonatomic) UIStackView *stackView;
+@property (nonatomic) NSArray<BPKStar *> *stars;
+@property (nonatomic) NSUInteger maxRating;
+
+@end
+
+@implementation BPKStarRating
+
+- (instancetype)initWithSize:(BPKStarSize)size {
+    BPKAssertMainThread();
+    self = [super initWithFrame:CGRectZero];
+    
+    if (self) {
+        [self setupWithSize:size];
+    }
+    
+    return self;
+}
+
+- (instancetype)initWithFrame:(CGRect)frame {
+    BPKAssertMainThread();
+    self = [super initWithFrame:frame];
+    
+    if (self) {
+        [self setupWithSize:BPKStarSizeSmall];
+    }
+    
+    return self;
+}
+
+- (nullable instancetype)initWithCoder:(NSCoder *)aDecoder {
+    BPKAssertMainThread();
+    self = [super initWithCoder:aDecoder];
+
+    if (self) {
+        [self setupWithSize:BPKStarSizeSmall];
+    }
+
+    return self;
+}
+
+- (void)setupWithSize:(BPKStarSize)size {
+    _maxRating = 5;
+    _size = size;
+    _rating = 0.0;
+    [self setupStackView];
+    [self setupStars];
+}
+
+- (void)setupStackView {
+    _stackView = [[UIStackView alloc] initWithFrame:CGRectZero];
+    _stackView.translatesAutoresizingMaskIntoConstraints = NO;
+    [self addSubview:_stackView];
+    
+    [NSLayoutConstraint activateConstraints:@[[self.stackView.topAnchor constraintEqualToAnchor:self.topAnchor],
+                                              [self.stackView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
+                                              [self.stackView.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
+                                              [self.stackView.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
+                                              ]];
+}
+
+- (void)setupStars {
+    NSMutableArray *temp = [NSMutableArray array];
+    
+    for (int i = 0; i < self.maxRating; i++) {
+        BPKStar *starView = [[BPKStar alloc] initWithSize:self.size];
+        [_stackView addArrangedSubview:starView];
+        [temp addObject:starView];
+    }
+    
+    _stars = [temp copy];
+}
+
+#pragma mark - Layout
+
+- (CGSize)intrinsicContentSize {
+    return [self.stackView intrinsicContentSize];
+}
+
+#pragma mark - Properties
+
+- (void)setRating:(float)rating {
+    if (_rating != rating) {
+        _rating = rating;
+        [self updateRating];
+    }
+}
+
+- (void)setSize:(BPKStarSize)size {
+    if (_size != size) {
+        _size = size;
+        [self updateSize];
+    }
+}
+
+#pragma mark - Updates
+
+- (void)updateRating {
+    float roundedRating = round(self.rating * 2.0) / 2.0;
+    for (NSInteger i = 0; i < self.stars.count; i++) {
+        if (i + 1 <= roundedRating) {
+            self.stars[i].state = BPKStarStateFull;
+        } else if (i + 0.5 == roundedRating) {
```

Comment:
This needs to rewritten to account for floating point percision problems. 

```swift
print("0.1 + 0.2 == 0.3 = \(0.1 + 0.2 == 0.3)")
```

[See Repl](https://repl.it/repls/SpottedExperiencedTransversals)

Reply:
Done